File: src/my_robot_object_detection/my_robot_object_detection/object_detector.py
#!/usr/bin/env python3
"""
Object Detection Node using OpenCV and ROS2
Detects colored objects (default: red objects) in camera feed
Publishes annotated images and detection results
"""

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from geometry_msgs.msg import Point
from std_msgs.msg import Bool
from cv_bridge import CvBridge
import cv2
import numpy as np


class ObjectDetector(Node):
    """ROS2 node for detecting colored objects using OpenCV"""

    def __init__(self):
        super().__init__('object_detector')

        # Declare parameters for color detection (HSV range for red object)
        self.declare_parameter('lower_hue_1', 0)
        self.declare_parameter('lower_saturation_1', 100)
        self.declare_parameter('lower_value_1', 100)
        self.declare_parameter('upper_hue_1', 10)
        self.declare_parameter('upper_saturation_1', 255)
        self.declare_parameter('upper_value_1', 255)

        # Second range for red (wraps around in HSV)
        self.declare_parameter('lower_hue_2', 170)
        self.declare_parameter('lower_saturation_2', 100)
        self.declare_parameter('lower_value_2', 100)
        self.declare_parameter('upper_hue_2', 180)
        self.declare_parameter('upper_saturation_2', 255)
        self.declare_parameter('upper_value_2', 255)

        # Detection parameters
        self.declare_parameter('min_area', 500)
        self.declare_parameter('camera_topic', '/my_robot/camera/image_raw')
        self.declare_parameter('show_debug', True)

        # Get parameters
        self.lower_hsv_1 = np.array([
            self.get_parameter('lower_hue_1').value,
            self.get_parameter('lower_saturation_1').value,
            self.get_parameter('lower_value_1').value
        ])
        self.upper_hsv_1 = np.array([
            self.get_parameter('upper_hue_1').value,
            self.get_parameter('upper_saturation_1').value,
            self.get_parameter('upper_value_1').value
        ])
        self.lower_hsv_2 = np.array([
            self.get_parameter('lower_hue_2').value,
            self.get_parameter('lower_saturation_2').value,
            self.get_parameter('lower_value_2').value
        ])
        self.upper_hsv_2 = np.array([
            self.get_parameter('upper_hue_2').value,
            self.get_parameter('upper_saturation_2').value,
            self.get_parameter('upper_value_2').value
        ])
        self.min_area = self.get_parameter('min_area').value
        camera_topic = self.get_parameter('camera_topic').value
        self.show_debug = self.get_parameter('show_debug').value

        # Initialize CV Bridge
        self.bridge = CvBridge()

        # Publishers
        self.image_pub = self.create_publisher(
            Image,
            '/object_detection/image_annotated',
            10
        )
        self.detection_pub = self.create_publisher(
            Point,
            '/object_detection/center',
            10
        )
        self.detected_pub = self.create_publisher(
            Bool,
            '/object_detection/detected',
            10
        )

        # Subscriber to camera feed
        self.image_sub = self.create_subscription(
            Image,
            camera_topic,
            self.image_callback,
            10
        )

        self.get_logger().info('Object Detector Node Initialized')
        self.get_logger().info(f'Subscribed to: {camera_topic}')
        self.get_logger().info(f'HSV Range 1: {self.lower_hsv_1} to {self.upper_hsv_1}')
        self.get_logger().info(f'HSV Range 2: {self.lower_hsv_2} to {self.upper_hsv_2}')
        self.get_logger().info(f'Minimum detection area: {self.min_area} pixels')

    def image_callback(self, msg):
        """Process incoming camera images and detect objects"""
        try:
            # Convert ROS Image message to OpenCV image
            cv_image = self.bridge.imgmsg_to_cv2(msg, desired_encoding='bgr8')

            # Convert BGR to HSV color space
            hsv = cv2.cvtColor(cv_image, cv2.COLOR_BGR2HSV)

            # Create masks for color detection (red has two ranges in HSV)
            mask1 = cv2.inRange(hsv, self.lower_hsv_1, self.upper_hsv_1)
            mask2 = cv2.inRange(hsv, self.lower_hsv_2, self.upper_hsv_2)
            mask = cv2.bitwise_or(mask1, mask2)

            # Morphological operations to remove noise
            kernel = np.ones((5, 5), np.uint8)
            mask = cv2.erode(mask, kernel, iterations=2)
            mask = cv2.dilate(mask, kernel, iterations=2)

            # Find contours
            contours, _ = cv2.findContours(
                mask,
                cv2.RETR_EXTERNAL,
                cv2.CONTOUR_APPROX_SIMPLE
            )

            # Create annotated image
            annotated_image = cv_image.copy()
            object_detected = False
            center_point = Point()

            # Process contours
            if contours:
                # Find the largest contour
                largest_contour = max(contours, key=cv2.contourArea)
                area = cv2.contourArea(largest_contour)

                if area > self.min_area:
                    object_detected = True

                    # Get bounding rectangle
                    x, y, w, h = cv2.boundingRect(largest_contour)

                    # Calculate center
                    cx = x + w // 2
                    cy = y + h // 2

                    # Store center point
                    center_point.x = float(cx)
                    center_point.y = float(cy)
                    center_point.z = float(area)  # Store area in z coordinate

                    # Draw bounding box
                    cv2.rectangle(
                        annotated_image,
                        (x, y),
                        (x + w, y + h),
                        (0, 255, 0),
                        3
                    )

                    # Draw center point
                    cv2.circle(
                        annotated_image,
                        (cx, cy),
                        5,
                        (0, 0, 255),
                        -1
                    )

                    # Draw crosshair
                    cv2.line(
                        annotated_image,
                        (cx - 20, cy),
                        (cx + 20, cy),
                        (0, 0, 255),
                        2
                    )
                    cv2.line(
                        annotated_image,
                        (cx, cy - 20),
                        (cx, cy + 20),
                        (0, 0, 255),
                        2
                    )

                    # Add text information
                    info_text = f"Object Detected | Area: {int(area)}"
                    cv2.putText(
                        annotated_image,
                        info_text,
                        (10, 30),
                        cv2.FONT_HERSHEY_SIMPLEX,
                        0.7,
                        (0, 255, 0),
                        2
                    )

                    position_text = f"Center: ({cx}, {cy})"
                    cv2.putText(
                        annotated_image,
                        position_text,
                        (10, 60),
                        cv2.FONT_HERSHEY_SIMPLEX,
                        0.7,
                        (0, 255, 0),
                        2
                    )

                    if self.show_debug:
                        self.get_logger().info(
                            f'Object detected at ({cx}, {cy}) with area {int(area)}'
                        )

            # Add status text if no object detected
            if not object_detected:
                cv2.putText(
                    annotated_image,
                    "No Object Detected",
                    (10, 30),
                    cv2.FONT_HERSHEY_SIMPLEX,
                    0.7,
                    (0, 0, 255),
                    2
                )

            # Publish results
            try:
                annotated_msg = self.bridge.cv2_to_imgmsg(
                    annotated_image,
                    encoding='bgr8'
                )
                self.image_pub.publish(annotated_msg)
            except Exception as e:
                self.get_logger().error(f'Failed to publish annotated image: {e}')

            # Publish detection status
            detected_msg = Bool()
            detected_msg.data = object_detected
            self.detected_pub.publish(detected_msg)

            # Publish center point if object detected
            if object_detected:
                self.detection_pub.publish(center_point)

        except Exception as e:
            self.get_logger().error(f'Error processing image: {e}')


def main(args=None):
    rclpy.init(args=args)
    node = ObjectDetector()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/my_robot_object_detection/src/object_detector.cpp
/**
 * @file object_detector.cpp
 * @brief Object Detection Node using OpenCV and ROS2
 *
 * Detects colored objects (default: red objects) in camera feed
 * Publishes annotated images and detection results
 */

#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>
#include <geometry_msgs/msg/point.hpp>
#include <std_msgs/msg/bool.hpp>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>
#include <opencv2/imgproc/imgproc.hpp>

class ObjectDetector : public rclcpp::Node
{
public:
    ObjectDetector() : Node("object_detector")
    {
        // Declare parameters for color detection (HSV range for red object)
        this->declare_parameter("lower_hue_1", 0);
        this->declare_parameter("lower_saturation_1", 100);
        this->declare_parameter("lower_value_1", 100);
        this->declare_parameter("upper_hue_1", 10);
        this->declare_parameter("upper_saturation_1", 255);
        this->declare_parameter("upper_value_1", 255);

        // Second range for red (wraps around in HSV)
        this->declare_parameter("lower_hue_2", 170);
        this->declare_parameter("lower_saturation_2", 100);
        this->declare_parameter("lower_value_2", 100);
        this->declare_parameter("upper_hue_2", 180);
        this->declare_parameter("upper_saturation_2", 255);
        this->declare_parameter("upper_value_2", 255);

        // Detection parameters
        this->declare_parameter("min_area", 500);
        this->declare_parameter("camera_topic", "/my_robot/camera/image_raw");
        this->declare_parameter("show_debug", true);

        // Get parameters
        lower_hsv_1_ = cv::Scalar(
            this->get_parameter("lower_hue_1").as_int(),
            this->get_parameter("lower_saturation_1").as_int(),
            this->get_parameter("lower_value_1").as_int()
        );
        upper_hsv_1_ = cv::Scalar(
            this->get_parameter("upper_hue_1").as_int(),
            this->get_parameter("upper_saturation_1").as_int(),
            this->get_parameter("upper_value_1").as_int()
        );
        lower_hsv_2_ = cv::Scalar(
            this->get_parameter("lower_hue_2").as_int(),
            this->get_parameter("lower_saturation_2").as_int(),
            this->get_parameter("lower_value_2").as_int()
        );
        upper_hsv_2_ = cv::Scalar(
            this->get_parameter("upper_hue_2").as_int(),
            this->get_parameter("upper_saturation_2").as_int(),
            this->get_parameter("upper_value_2").as_int()
        );
        min_area_ = this->get_parameter("min_area").as_int();
        show_debug_ = this->get_parameter("show_debug").as_bool();
        std::string camera_topic = this->get_parameter("camera_topic").as_string();

        // Publishers
        image_pub_ = this->create_publisher<sensor_msgs::msg::Image>(
            "/object_detection/image_annotated", 10);
        detection_pub_ = this->create_publisher<geometry_msgs::msg::Point>(
            "/object_detection/center", 10);
        detected_pub_ = this->create_publisher<std_msgs::msg::Bool>(
            "/object_detection/detected", 10);

        // Subscriber to camera feed
        image_sub_ = this->create_subscription<sensor_msgs::msg::Image>(
            camera_topic, 10,
            std::bind(&ObjectDetector::imageCallback, this, std::placeholders::_1));

        RCLCPP_INFO(this->get_logger(), "Object Detector Node Initialized");
        RCLCPP_INFO(this->get_logger(), "Subscribed to: %s", camera_topic.c_str());
        RCLCPP_INFO(this->get_logger(), "HSV Range 1: [%d, %d, %d] to [%d, %d, %d]",
            (int)lower_hsv_1_[0], (int)lower_hsv_1_[1], (int)lower_hsv_1_[2],
            (int)upper_hsv_1_[0], (int)upper_hsv_1_[1], (int)upper_hsv_1_[2]);
        RCLCPP_INFO(this->get_logger(), "HSV Range 2: [%d, %d, %d] to [%d, %d, %d]",
            (int)lower_hsv_2_[0], (int)lower_hsv_2_[1], (int)lower_hsv_2_[2],
            (int)upper_hsv_2_[0], (int)upper_hsv_2_[1], (int)upper_hsv_2_[2]);
        RCLCPP_INFO(this->get_logger(), "Minimum detection area: %d pixels", min_area_);
    }

private:
    void imageCallback(const sensor_msgs::msg::Image::SharedPtr msg)
    {
        try {
            // Convert ROS Image message to OpenCV image
            cv_bridge::CvImagePtr cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
            cv::Mat cv_image = cv_ptr->image;

            // Convert BGR to HSV color space
            cv::Mat hsv;
            cv::cvtColor(cv_image, hsv, cv::COLOR_BGR2HSV);

            // Create masks for color detection (red has two ranges in HSV)
            cv::Mat mask1, mask2, mask;
            cv::inRange(hsv, lower_hsv_1_, upper_hsv_1_, mask1);
            cv::inRange(hsv, lower_hsv_2_, upper_hsv_2_, mask2);
            cv::bitwise_or(mask1, mask2, mask);

            // Morphological operations to remove noise
            cv::Mat kernel = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(5, 5));
            cv::erode(mask, mask, kernel, cv::Point(-1, -1), 2);
            cv::dilate(mask, mask, kernel, cv::Point(-1, -1), 2);

            // Find contours
            std::vector<std::vector<cv::Point>> contours;
            std::vector<cv::Vec4i> hierarchy;
            cv::findContours(mask, contours, hierarchy, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);

            // Create annotated image
            cv::Mat annotated_image = cv_image.clone();
            bool object_detected = false;
            geometry_msgs::msg::Point center_point;

            // Process contours
            if (!contours.empty()) {
                // Find the largest contour
                auto largest_contour = std::max_element(
                    contours.begin(), contours.end(),
                    [](const std::vector<cv::Point>& c1, const std::vector<cv::Point>& c2) {
                        return cv::contourArea(c1) < cv::contourArea(c2);
                    }
                );

                double area = cv::contourArea(*largest_contour);

                if (area > min_area_) {
                    object_detected = true;

                    // Get bounding rectangle
                    cv::Rect bounding_rect = cv::boundingRect(*largest_contour);
                    int x = bounding_rect.x;
                    int y = bounding_rect.y;
                    int w = bounding_rect.width;
                    int h = bounding_rect.height;

                    // Calculate center
                    int cx = x + w / 2;
                    int cy = y + h / 2;

                    // Store center point
                    center_point.x = static_cast<double>(cx);
                    center_point.y = static_cast<double>(cy);
                    center_point.z = area;  // Store area in z coordinate

                    // Draw bounding box
                    cv::rectangle(annotated_image,
                                cv::Point(x, y),
                                cv::Point(x + w, y + h),
                                cv::Scalar(0, 255, 0), 3);

                    // Draw center point
                    cv::circle(annotated_image, cv::Point(cx, cy), 5, cv::Scalar(0, 0, 255), -1);

                    // Draw crosshair
                    cv::line(annotated_image,
                           cv::Point(cx - 20, cy),
                           cv::Point(cx + 20, cy),
                           cv::Scalar(0, 0, 255), 2);
                    cv::line(annotated_image,
                           cv::Point(cx, cy - 20),
                           cv::Point(cx, cy + 20),
                           cv::Scalar(0, 0, 255), 2);

                    // Add text information
                    std::string info_text = "Object Detected | Area: " + std::to_string(static_cast<int>(area));
                    cv::putText(annotated_image, info_text,
                              cv::Point(10, 30), cv::FONT_HERSHEY_SIMPLEX,
                              0.7, cv::Scalar(0, 255, 0), 2);

                    std::string position_text = "Center: (" + std::to_string(cx) + ", " + std::to_string(cy) + ")";
                    cv::putText(annotated_image, position_text,
                              cv::Point(10, 60), cv::FONT_HERSHEY_SIMPLEX,
                              0.7, cv::Scalar(0, 255, 0), 2);

                    if (show_debug_) {
                        RCLCPP_INFO_THROTTLE(this->get_logger(), *this->get_clock(), 1000,
                            "Object detected at (%d, %d) with area %.0f", cx, cy, area);
                    }
                }
            }

            // Add status text if no object detected
            if (!object_detected) {
                cv::putText(annotated_image, "No Object Detected",
                          cv::Point(10, 30), cv::FONT_HERSHEY_SIMPLEX,
                          0.7, cv::Scalar(0, 0, 255), 2);
            }

            // Publish annotated image
            try {
                sensor_msgs::msg::Image::SharedPtr annotated_msg =
                    cv_bridge::CvImage(msg->header, "bgr8", annotated_image).toImageMsg();
                image_pub_->publish(*annotated_msg);
            } catch (cv_bridge::Exception& e) {
                RCLCPP_ERROR(this->get_logger(), "Failed to publish annotated image: %s", e.what());
            }

            // Publish detection status
            std_msgs::msg::Bool detected_msg;
            detected_msg.data = object_detected;
            detected_pub_->publish(detected_msg);

            // Publish center point if object detected
            if (object_detected) {
                detection_pub_->publish(center_point);
            }

        } catch (cv_bridge::Exception& e) {
            RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
        } catch (cv::Exception& e) {
            RCLCPP_ERROR(this->get_logger(), "OpenCV exception: %s", e.what());
        }
    }

    // Parameters
    cv::Scalar lower_hsv_1_;
    cv::Scalar upper_hsv_1_;
    cv::Scalar lower_hsv_2_;
    cv::Scalar upper_hsv_2_;
    int min_area_;
    bool show_debug_;

    // ROS2 publishers and subscribers
    rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr image_pub_;
    rclcpp::Publisher<geometry_msgs::msg::Point>::SharedPtr detection_pub_;
    rclcpp::Publisher<std_msgs::msg::Bool>::SharedPtr detected_pub_;
    rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr image_sub_;
};

int main(int argc, char** argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<ObjectDetector>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
